File: Java/org.tec.datastructures/src/org/tec/datastructures/hierarchical/BinaryTree.java
package org.tec.datastructures.hierarchical;

import org.tec.datastructures.nodes.BinaryTreeNode;

public class BinaryTree<T extends Comparable<T>> {
	
	private BinaryTreeNode<T> root;
	
	public BinaryTree() {
		this(null); 
	}
	
	public BinaryTree(BinaryTreeNode<T> Root) {
		root = Root;
	}
	
	public BinaryTreeNode<T> getRoot() {
		return root;
	}

	public void setRoot(BinaryTreeNode<T> root) {
		this.root = root;
	}

	public void append(T Value) {
		
		if (root == null) {
			root = new BinaryTreeNode<T>(Value);
			return;
		}
		root = appendRecursive(Value,root);
		
	}

	private BinaryTreeNode<T> appendRecursive(T value, BinaryTreeNode<T> node) {
		if (node == null) {
			node = new BinaryTreeNode<T>(value);
		}
		else if (node.getValue().compareTo(value) > 0) {
			node.setLeft(appendRecursive(value, node.getLeft()));
		}
		else if (node.getValue().compareTo(value) < 0) {
			node.setRight(appendRecursive(value, node.getRight()));
		}
		return node;
	}
	
	public BinaryTreeNode<T> search(T value) {
		return searchRecursive(value,root);
	}

	private BinaryTreeNode<T> searchRecursive(T value, BinaryTreeNode<T> node) {
		if (node == null) {
			return null;
		}
		else if (node.getValue().compareTo(value) > 0) {
			return searchRecursive(value, node.getLeft());
		}
		else if (node.getValue().compareTo(value) < 0) {
			return searchRecursive(value, node.getRight());
		}
		return node;
	}
	
	public void delete(T value) {
		root = deleteRecursive(value, root);
	}
	
	private BinaryTreeNode<T> deleteRecursive(T value, BinaryTreeNode<T> node) {
		if (node == null) {
			return null;
		}
		else if (node.getValue().compareTo(value) > 0) {
			node.setLeft(deleteRecursive(value, node.getLeft()));
		}
		else if (node.getValue().compareTo(value) < 0) {
			node.setRight(deleteRecursive(value, node.getRight()));
		} 
		else if (node.getLeft() != null && node.getRight() != null) {
			T replace = findMin(node.getRight()).getValue();
			
			node.setValue(replace);
			node.setRight(deleteRecursive(replace,node.getRight()));
		}
		else {
			node = (node.getLeft() == null) ? node.getRight() : node.getLeft();
		}
		return node;
	}

	private BinaryTreeNode<T> findMin(BinaryTreeNode<T> node) {
		if (node.getLeft() == null) {
			return node;
		}
		return findMin(node.getLeft());
	}
	
}

File: C#/org.tec.datastructures2/BinaryTree.cs
using System;
using System.Collections;
namespace org.tec.datastructures
{
    public class BinaryTree<T> where T : IComparable<T>
    {
        private BinaryNode<T> root = null;

        public void Append(T value){
            root = AppendAux(root, value);
        }

        private BinaryNode<T> AppendAux(BinaryNode<T> node, T value){
            if (node == null){
                node = new BinaryNode<T>(value);
            }else if(node.value.CompareTo(value) == 1){
                node.left = AppendAux(node.left, value);
            }else if(node.value.CompareTo(value) == -1){
                node.right = AppendAux(node.right, value);
            }
            return node;
        }

        public BinaryNode<T> Get(T value){
            return GetAux(root, value);
        }

        private BinaryNode<T> GetAux(BinaryNode<T> node, T value){
            if (node == null){}
            else if (node.value.CompareTo(value) == 1)
            {
                return GetAux(node.left, value);
            }
            else if (node.value.CompareTo(value) == -1)
            {
                return GetAux(node.right, value);
            }
            return node;
        }

        public void Delete(T value)
        {
            root = DeleteAux(root, value);
        }

        private BinaryNode<T> DeleteAux(BinaryNode<T> node, T value)
        {
            if (node == null){}
            else if (node.value.CompareTo(value) > 0)
            {
                node.left = DeleteAux(node.left, value);
            }
            else if (node.value.CompareTo(value) < 0)
            {
                node.right = DeleteAux(node.right, value);
            }
            else if (node.left != null && node.right != null)
            {
                T replace = FindMin(node.right).value;

                node.value = replace;
                node.right = DeleteAux(node.right, value);
            }
            else
            {
                node = node.left == null ? node.right : node.left;
            }
            return node;
        }

        private BinaryNode<T> FindMin(BinaryNode<T> node)
        {
            if (node.left == null)
            {
                return node;
            }
            return FindMin(node.left);
        }
    }

    public class BinaryNode<T>{
        public BinaryNode<T> left = null;
        public BinaryNode<T> right = null;
        public T value;

        public BinaryNode(T value){
            this.value = value;
        }

        public Boolean hasChilden(){
            return left != null || right != null;
        }
    }


}
